### PR TITLE
Add FileReaderBuilder for arrow-ipc to allow reading large no. of column files

### LINKED ISSUE #4432: Implicit one million column limit on arrow files
**Describe the bug**

Attempting to construct an `arrow::ipc::reader::FileReader` using `FileReader::try_new` on an arrow file with more than one million columns fails.

**To Reproduce**

Running the following code on a suitable file with greater than one million columns,
```rust
let mut file = File::open("my-file-with-gt-1million-cols.arrow").unwrap();
let reader = FileReader::try_new(reader, None).unwrap(); // arrow::ipc::reader::FileReader
```
Produces the following error:
```
panicked at 'called `Result::unwrap()` on an `Err` value: IoError("Unable to get root as footer: TooManyTables")'
```

**Expected behavior**

You would expect the `FileReader` to construct normally.

**Additional context**

I took some time to understand this bug and it's upstream from a flatbuffer default setting. I believe this is a direct consequence of this chain of calls: FileReader::try_new -> File::root_as_footer -> flatbuffers::root, which constructs a VerifierOptions with default arguments for which VerifierOptions::default().max_tables is 1,000,000.  The result of all of this is when you call FileReader::try_new on a file that contains > 1,000,000 columns, you get an error message propagating from [this line](https://docs.rs/arrow-ipc/41.0.0/src/arrow_ipc/reader.rs.html#691) which is just upstream from [here](https://github.com/google/flatbuffers/blob/23922e7eba51e666f9af13942ddb6cd6d58ef0de/rust/flatbuffers/src/verifier.rs#L418). 

I believe this issue could be fixed by changing [`root_as_footer`](https://github.com/apache/arrow-rs/blob/15e0e76bfb6500799a43991f1339e69464c513f8/arrow-ipc/src/gen/File.rs#LL444C5-L444C5) to first construct a `VerifierOptions` with a different option for `max_tables` and then pass it to `flatbuffers::root_with_opts` along with the buffer. This gets around the default one million limit imposed by [`flatbuffers::root`](https://github.com/google/flatbuffers/blob/23922e7eba51e666f9af13942ddb6cd6d58ef0de/rust/flatbuffers/src/get_root.rs#L26) which just constructs a `VerifierOptions::default()` and passes it to `root_with_opts` anyway.

Additionally, `arrow2` seems to have [dealt with this issue previously](https://github.com/jorgecarleitao/arrow2/commit/8e146a78f53899dd7f7166512c1b73a51803b0ab).

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

Closes #4432

# Rationale for this change
 

Following suggestion as per https://github.com/apache/arrow-rs/pull/4434#discussion_r1242061745

# What changes are included in this PR?

New `FileReaderBuilder` for arrow-ipc reader to allow configuring the `VerifierOptions` which can allow reading an IPC file with massive number of columns (over 1 million) if user configures correctly.

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-ipc/src/reader.rs [hunk 1]
```diff
@@ -498,61 +498,39 @@ pub fn read_dictionary(
...
-            .field("dictionaries_by_id", &self.dictionaries_by_id)
-            .field("metadata_version", &self.metadata_version)
-            .field("projection", &self.projection)
-            .finish()
+    /// Flatbuffers options for parsing footer. Useful if needing to parse a file containing
+    /// millions of columns, in which case can up the value for `max_tables` to accommodate parsing
+    /// such a file.
+    pub fn with_verifier_options(mut self, verifier_options: VerifierOptions) -> Self {
+        self.verifier_options = verifier_options;
+        self
```
Comment (Jefffrey) [inline arrow-ipc/src/reader.rs:529]: Wasn't sure if preferable to have this for maximum flexibility, or have something as suggested here: https://github.com/apache/arrow-rs/pull/4434#issuecomment-1615121403

Which might be more user friendly, though abstracts away the inner flatbuffers setting being changed.

**Comment by Jefffrey** [comment]
I've refactored to not expose the flatbuffer types, but am still keeping the flatbuffer terminology. I figured that since tables refers to both key-value metadata pairs and also columns, it would be better to just expose the flatbuffer setting and document it rather than name it `with_approx_max_columns` or something similar as that could cause confusion since it'll also affect/be affected by number of metadata key-value pairs

Also added setting for depth while I was at it

**Review by tustvold — APPROVED** [review]
Thank you

**Comment by tustvold** [comment]
I have double-checked that it is only the footer where we need to handle this, as the message encoding is already flattened and doesn't contain nested tables.

**Pull request merged**